docs/developer-guide/creating-agents.md

```diff
@@ -0,0 +1,763 @@
+        
+        return AgentOutput(
+            agent_name=self.name,
+            content=response,
+            metadata={
+                "api_endpoint": f"{self.base_url}/analyze",
+                "api_response_time": response.get("processing_time", 0)
+            }
+        )
+    
+    async def _call_api(self, content: str) -> dict:
+        """Call external API."""
+        headers = {"Authorization": f"Bearer {self.api_key}"}
+        payload = {"text": content}
+        
+        async with self.session.post(
+            f"{self.base_url}/analyze",
+            json=payload,
+            headers=headers
+        ) as response:
+            return await response.json()
+```
+
+## Testing Agents
+
+### Unit Tests
+
+```python
+import pytest
+from unittest.mock import Mock, patch
+from buttermilk._core.types import Record
+
+class TestContentAnalyzer:
+    @pytest.fixture
+    def agent(self):
+        return ContentAnalyzer(
+            agent_id="test_agent",
+            role="ANALYZER",
+            description="Test analyzer",
+            analysis_type="sentiment"
+        )
+    
+    @pytest.fixture
+    def sample_record(self):
+        return Record(
+            id="test_1",
+            content="This is a test message with positive sentiment."
+        )
+    
+    async def test_agent_processes_record(self, agent, sample_record):
+        # Mock LLM client
+        with patch.object(agent, 'llm_client') as mock_llm:
+            mock_llm.generate.return_value = "positive"
+            
+            result = await agent._process(record=sample_record)
+            
+            assert result.agent_name == agent.name
+            assert result.content == "positive"
+            assert result.metadata["analysis_type"] == "sentiment"
+    
+    async def test_agent_handles_error(self, agent, sample_record):
+        with patch.object(agent, 'llm_client') as mock_llm:
+            mock_llm.generate.side_effect = Exception("API Error")
+            
+            with pytest.raises(Exception, match="API Error"):
+                await agent._process(record=sample_record)
+```
+
+### Integration Tests
+
+```python
+@pytest.mark.asyncio
+async def test_agent_with_real_llm():
+    agent = ContentAnalyzer(
+        agent_id="integration_test",
+        role="ANALYZER",
+        description="Integration test",
+        analysis_type="sentiment",
+        model="gemini-pro"
+    )
+    
+    record = Record(
+        id="test_1",
+        content="I love this new feature!"
+    )
+    
+    result = await agent._process(record=record)
+    
+    assert result.success
+    assert result.content
+    assert "positive" in result.content.lower()
+```
+
+## Best Practices
+
+### 1. Agent Design Principles
+
+**Single Responsibility:**
+- Each agent should have one clear purpose
+- Avoid creating agents that do too many things
+
+**Stateless Processing:**
+- Agents should be stateless by default
```

Comment:
incorrect. agents in a flow have state.

Reply:
but maybe they shouldn't? some need context, others don't. let's add this as an issue for discussion.